Implement BLS12-381 `contracttype` support

jayz22 opened the pull request:
### What

Add support for BLS12-381 curve points (`Fp`, `Fp2`, `G1Affine`, `G2Affine`, `Fr`) for internal data storage (via `contracttype`) and contract invocation arguments (contract spec). 

### Why

Improves usability of the BLS12-381 features in a Groth16 verifier application ([example contract](https://github.com/stellar/soroban-examples/pull/350))
-  `Arbitrary` and several missing conversions are needed to use `G1Affine`, `G2Affine` as contract data.

### Known limitations

[TODO or N/A]

jayz22 committed:
implement missing traits for G1/G2/Fr to be used as `contracttype`

Add comment for clearing flags, remove dead code

properly handle infinity bit in the `Arbitrary` type generation

leighmcculloch commented on soroban-sdk/src/crypto/bls12_381.rs:
```diff
@@ -115,6 +217,14 @@ impl G1Affine {
     }
 }
 
+// impl TryFromVal<Env, Val> for G1Affine {
+//     type Error = ConversionError;
+
+//     fn try_from_val(env: &Env, v: &Val) -> Result<Self, Self::Error> {
+//         Ok(G1Affine(BytesN::try_from_val(env, v)?))
+//     }
+// }
```
?

jayz22 replied:
removed

leighmcculloch commented on soroban-sdk/src/crypto/bls12_381.rs:
```diff
@@ -7,15 +7,66 @@ use crate::{
+
+impl<const N: usize, const M: usize> Into<BigInt<N>> for BytesN<M> {
+    fn into(self) -> BigInt<N> {
+        if M != N * 8 {
+            sdk_panic!("BytesN::Into<BigInt> - length mismatch")
+        }
```
The `M != N*8` check occurs at runtime. We should perform this check at compile time so that devs find out earlier in the dev cycle that they've written code that will only fail.

Since [Rust 1.79](https://doc.rust-lang.org/beta/releases.html#version-1790-2024-06-13) we can do that using the `const { /* ... */ }` expression. The SDK requires at least Rust 1.79, so it is safe to use it in the SDK.

The const block will cause everything in it to be evaluated at build time, and the panic will cause a compile error.

```suggestion
        const {
            if M != N * 8 {
                sdk_panic!("BytesN::Into<BigInt> - length mismatch")
            }
        }
```

The other change we should make is that `sdk_panic!` is only useful in runtime contexts. The macro makes sure that the text given in the panic message is not stored in the wasm for use at runtime. In a const / compile time context using `panic!` has no cost, and is better, because it will mean that even when building with a wasm target will the panic text be outputted in the compile error.

So this is what I'd suggest.

```suggestion
        const {
            if M != N * 8 {
                panic!("BytesN::Into<BigInt> - length mismatch")
            }
        }
```

We should use this pattern in the other places where the values of N and M are validated.

jayz22 replied:
great suggestion, and good to know about the const expression feature! 
fixed it in the other PR [here](https://github.com/stellar/rs-soroban-sdk/pull/1456#discussion_r2026154467) and [here](https://github.com/stellar/rs-soroban-sdk/pull/1456#discussion_r2026155118)

leighmcculloch commented on soroban-sdk/src/crypto/bls12_381.rs:
```diff
@@ -26,7 +77,7 @@ pub struct Bls12_381 {
 ///   `Fp`
 /// - The most significant three bits (bits 0-3) of the first byte are reserved
 ///   for encoding flags:
-///   - compression_flag (bit 0): Must always be set (1), as only uncompressed
+///   - compression_flag (bit 0): Must always be unset (0), as only uncompressed
```
Is this behaviour that's changing that's related to this change, or a bug in documentation? If latter, and since this looks isolated from everything else here, could you make this fix in a separate PR? so that it is captured clearly in the changelog as its own line item, and in general to increase the visibility.

jayz22 replied:
it is a bug in the documentation. separated out the fix in https://github.com/stellar/rs-soroban-sdk/pull/1455

leighmcculloch commented on soroban-sdk/src/crypto/bls12_381.rs:
```diff
@@ -101,6 +152,57 @@ impl_bytesn_repr!(G2Affine, 192);
+        self.0.env()
+    }
+
+    pub fn map_to_g1(&self) -> G1Affine {
```
It looks like this method converts an `Fp` to a `G1Affine`. Is that accurate? Or is there semantics about the `map` part of it that represent more than a conversion from one type to the next?

Rust naming conventions suggest using `into`, `as_`, and `to_` for type conversions, so it'd be preferred to follow that naming convention unless this shouldn't be thought of as a conversion. The most appropriate for this one is probably `to_g1` because it's a borrowed to owned. See https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv.

```suggestion
    pub fn to_g1(&self) -> G1Affine {
```

jayz22 replied:
actually this is not a simple conversion, but performs mapping algorithms to map a number into a point. (the particular method is described [here](https://www.rfc-editor.org/rfc/rfc9380.html#name-simplified-swu-for-ab-0)). I've also added a comment in the code.

leighmcculloch commented on soroban-sdk/src/crypto/bls12_381.rs:
```diff
@@ -101,6 +152,57 @@ impl_bytesn_repr!(G2Affine, 192);
+    type Output = Fp;
+
+    fn neg(self) -> Self::Output {
+        if self.to_array() == [0; 48] {
+            return self;
+        }
+
+        let env = self.env().clone();
+        let fp_bigint: BigInt<6> = self.0.into();
```
`to_array` contains a host call to copy the memory to the guest, and when using `neg` that host call and copy is occurring twice, because it occurs at the beginning of the neg, and then inside the `Into<BigInt<6>>::into`.

It would be ideal to avoid the additional host call and host-guest memcpy.

You could add a function `zero` to `BigInt<6>` and do the zero check after converting into BigInt. The conversion for zero will be more expensive, but is zero usage an edge case? That tradeoff might be more appropriate if it is, so we don't hit the double memcpy for the general case.

jayz22 replied:
This is a great point. The `zero` checking is rather a corner case that isn't worth always paying the extra memcpy cost for. It should be sufficient to be done after the `Into::<BigInt<6>>` conversion, as you suggested. I've made the change in the other PR (https://github.com/stellar/rs-soroban-sdk/pull/1456#discussion_r2026156499)

leighmcculloch reviewed (commented):
Thanks for making these types usable with contracttype 👏🏻.

This pull request looks like it contains a bundle of changes related only by topic, and otherwise independent. It's helpful if it's split, both for changelog capturing, visibility, and readers. I understand that because the example (https://github.com/stellar/soroban-examples/pull/350) was being developed at the same time that also might not be desirable. No need to do now if too inconvenient, just fyi as the reader initially felt like things were related and had to piece through what was and wasn't, and intent for each change.

I've left a few comments inline to do with rust conventions, and changing a runtime check to a compile time check.

For the low-level BLS operations, I'm not the best person to review, so I'm tagging @stellar/contract-committers to review.

jayz22 commented:
@leighmcculloch I've split out this PR into three parts
1. https://github.com/stellar/rs-soroban-sdk/pull/1455 contains the comment fix
2. https://github.com/stellar/rs-soroban-sdk/pull/1456 contains the implementation of point negation
3. this PR just contains the `contracttype` support

they are currently based on top of one another, so they should be merged in the above order. I will mention and address your review comments pertaining to the those changes in the their corresponding PR.

jayz22 commented on soroban-sdk/src/testutils/arbitrary.rs:
```diff
@@ -680,6 +681,64 @@ mod objects {
+
+        fn try_from_val(env: &Env, v: &ArbitraryG1Affine) -> Result<Self, Self::Error> {
+            let mut bytes = v.bytes;
+            bytes[0] &= 0b0001_1111; // Clear the top 3 bits
```
added comments in the code.

jayz22 commented on soroban-sdk/src/testutils/arbitrary.rs:
```diff
@@ -680,6 +681,64 @@ mod objects {
+        type Error = ConversionError;
+
+        fn try_from_val(env: &Env, v: &ArbitraryFr) -> Result<Self, Self::Error> {
+            // Convert bytes to Fr via U256
+            Ok(Fr::from_bytes(BytesN::from_array(env, &v.bytes)))
```
that's right.

leighmcculloch reviewed (commented):
This looks good to me. I only have one concern, mentioned inline, and my concern may be naive.

Let's get https://github.com/stellar/rs-soroban-sdk/pull/1456 merged, and then once this PR is updated I'll approve this PR.

jayz22 commented on soroban-sdk/src/testutils/arbitrary.rs:
```diff
@@ -680,6 +681,76 @@ mod objects {
+            // the rest of the bytes must be zeros. The host will reject any
+            // invalid input. Clearing the flags here just to give it better
+            // chance of a valid input.
+            bytes[0] &= 0b0001_1111;
```
Yeah you're right. The infinity is a valid point that should be allowed to be generated.
Fixed the handling of infinity point by checking the flag and clearing the rest of the bits.

jayz22 commented on soroban-sdk/src/testutils/arbitrary.rs:
```diff
@@ -680,6 +681,76 @@ mod objects {
+            // the rest of the bytes must be zeros. The host will reject any
+            // invalid input. Clearing the flags here just to give it better
+            // chance of a valid input.
+            bytes[0] &= 0b0001_1111;
```
fixed

leighmcculloch reviewed (commented):
> Add support for BLS12-381 curve points (G1Affine, G2Affine, Fr) as contract data (via contracttype)

There's no test testing that we've got everything we need here for it to work. Could you add a test in the sdk tests directory to make sure we aren't missing something?

Is the goal only to be able to write and read these from storage? Or also to be able to pass it as parameters into the contract? For the latter, which I think makes sense, we need to hook up specs. Normally types that live in the SDK end up in the ScSpecType list.

The only other SDK types that use `contracttype` macros internally are for auth, and those conveniently don't need exporting of specs (we turn that off) because they aren't externally callable so this limitation and gap hasn't had an impact until now.

The way we did this in the token sdk was use the contracttype macro, and when you import the crate its spec gets reexported implicitly. But that won't work to do that here because everyone imports this SDK.

These types are all aliases for `BytesN<_>`. The other case we have for that is `Hash`, but again that was a case not callable.

🤔 This is a bit of a conundrum. 

We can merge this as is, it'll work for storage, but fall apart at the boundaries of the contract as a fn input or output and when other tools like the CLI try and pass the value.

We have a couple paths:
- Do nothing, maybe that's enough for now.
- Add to the spec XDR as a type
- Find a way for the specs to see it as a BytesN for the sake of specs https://github.com/stellar/rs-soroban-sdk/blob/13263e86813c3991e805f3d55666fa0402c1bf26/soroban-spec-rust/src/types.rs#L174-L177
- Explore other approaches

Related issues:
- https://github.com/stellar/rs-soroban-sdk/issues/1330

leighmcculloch commented:
> * Find a way for the specs to see it as a BytesN for the sake of specs

The way we'd do this is modify the `contractimpl` macro that generates the functions, so that for any input parameter named one of these types, we map them to the `BytesN<_>` type of appropriate size.

jayz22 committed:
Add BLS contract test

jayz22 commented on soroban-sdk/src/crypto/bls12_381.rs:
```diff
@@ -1,3 +1,5 @@
+#[cfg(not(target_family = "wasm"))]
+use crate::xdr::ScVal;
```
This is necessary because of the `impl_bytesn_repr` macro.

leighmcculloch reviewed (approved):
The test shows that we don't have everything we need here for these to work everywhere `contracttype`s do.

If I deploy the test and run this command, it errors:
```
$ stellar contract invoke --id c -- --help
❌ error: Missing Entry G1Affine
```

But there is enough here for these to work as internal types for storage. So I think we can merge this, because it is additive, and what's added works for that narrow purpose, but we should follow up on filling the gap on contract function params / return values.

leighmcculloch commented:
> The way we'd do this is modify the `contractimpl` macro that generates the functions, so that for any input parameter named one of these types, we map them to the `BytesN<_>` type of appropriate size.

If we solve this problem generically, not just hard code these types specifically, then I think we are also solving the type alias issue, if we also solve it for type wrapping:
- #1063

leighmcculloch commented:
To be clearer there are two contracttype issues outstanding:
- The SDK doesn't have a way to define a type that is optionally exported.
- The contract specs don't have a way to define type aliases, where one type is really another type. In this case it is the G1Affine newtype idiom around the BytesN type. This will be addressed by #1063.

In the interim of solving both problems above I think it is okay if the contracttype macro sees G1Affine and treats it as if the user had typed `BytesN<...>`. Doing that will avoid both problems for the moment.

jayz22 committed:
support bls types in contract specs

jayz22 committed:
fixup! support bls types in contract specs

jayz22 commented:
> To be clearer there are two contracttype issues outstanding:
> 
>     * The SDK doesn't have a way to define a type that is optionally exported, although the stellar-cli may do that for us in the future. [`stellar contract build` should strip contract specs, removing types that are not referenced by functions or events stellar-cli#2030](https://github.com/stellar/stellar-cli/issues/2030)
> 
>     * The contract specs don't have a way to define type aliases, where one type is really another type. In this case it is the G1Affine newtype idiom around the BytesN type. This will be addressed by [Support type aliases as contracttypes #1063](https://github.com/stellar/rs-soroban-sdk/issues/1063).
> 
> 
> In the interim, until both problems above are solved, I think it is okay if the `contractimpl` macro sees `G1Affine` and treats it as if the user had typed `BytesN<...>`. Doing that will avoid both problems for the moment.

Thanks @leighmcculloch, I've added the type spec mapping for the BLS types to its inner types (with a comment referencing #1063).
Also added a contract example and a test invoking it. I think that should do it for now. 
Agree it would be nice to support generic type inference (either via new type idiom or type alias).

jayz22 committed:
fixup! support bls types in contract specs

leighmcculloch committed:
update json

leighmcculloch committed:
add comments to clarify serialized size constants

leighmcculloch committed:
adjust bls12_381 constant formatting

leighmcculloch commented on soroban-sdk-macros/src/map_type.rs:
```diff
@@ -37,6 +40,31 @@ pub fn map_type(t: &Type, allow_hash: bool) -> Result<ScSpecTypeDef, Error> {
+                    // Idiom. For more details, see the tracking issue for supporting
+                    // type aliases:
+                    // https://github.com/stellar/rs-soroban-sdk/issues/1063
+                    "G1Affine" => Ok(ScSpecTypeDef::BytesN(ScSpecTypeBytesN {
+                        n: G1_SERIALIZED_SIZE,
+                    })),
+                    "G2Affine" => Ok(ScSpecTypeDef::BytesN(ScSpecTypeBytesN {
+                        n: G2_SERIALIZED_SIZE,
+                    })),
+                    "Fr" => Ok(ScSpecTypeDef::U256),
```
@jayz22 Do the `Fp` and `Fp2` types need listing here too?

jayz22 replied:
Hmm good call. My initial intuition was no since these are just field elements that makes up the coordinates for G1 and G2. No need to store/pass them as function argument directly. 
But looking at it again, we do have host functions `bls12_381_map_fp2/1_to_g2/1` that maps a field element to a point. 
So I guess let's just add it for completeness.

jayz22 replied:
Added Fp and Fp2 and updated the test contract to invoke host functions using those.

leighmcculloch committed:
disable the test snapshot file for the test using wasms

leighmcculloch committed:
fix

jayz22 committed:
add comment clarifying map_to_curve output's subgroup membership

jayz22 committed:
also support Fp, Fp2

leighmcculloch committed:
update bls12_381 documentation formatting using warning box

leighmcculloch commented:
I pushed a small change to make the warning stand out a little more, using rustdocs warning class:

Before:
<img width="654" alt="Screenshot 2025-05-16 at 1 44 16 pm" src="https://github.com/user-attachments/assets/486e497b-cce7-49e1-8b1a-9280f549098e" />

After:
<img width="654" alt="Screenshot 2025-05-16 at 1 43 56 pm" src="https://github.com/user-attachments/assets/e3209226-9caa-4c67-a878-f21de642389e" />

leighmcculloch committed:
bump stellar/binaries version and cargo-semver-checks
